### PR TITLE
fix: staleness detection now uses compare API instead of full PR diff

### PR DESCRIPTION
**Pull request by ichoosetoaccept**
## Summary

Replace the old blanket staleness detection (any file in PR diff = stale) with a precise per-thread temporal approach using the GitHub Compare API.

### What changed

- **New staleness model**: A thread is only stale if the file it comments on was modified by commits pushed *after* the comment's timestamp. Previously, any file in the PR diff would mark the thread stale — producing false positives when reviewers commented after a file was last changed.
- **Compare API**: `_compute_staleness` now uses `/repos/{owner}/{repo}/compare/{base}...{head}` to check which files changed between the review point and HEAD, instead of fetching the full PR diff.
- **`head_sha` derivation**: Uses the last commit in API topological order (`commits[-1]`), not the timestamp-sorted last entry — timestamps can be non-monotonic after rebase/cherry-pick.
- **Graceful error handling**: `_get_files_changed_between` catches `GhError` (e.g. 404 from force-pushed SHAs) and returns an empty set instead of killing the request.
- **Removed `_DIFF_QUERY`**: The old GraphQL diff query is no longer needed.

### Bug fixes

- **False-staleness edge case**: When a comment is after all commits by timestamp but `review_point_sha != head_sha` (non-monotonic timestamps), the old code would falsely mark it stale. Fixed by comparing `review_point_sha` against `timeline[-1]` instead of `head_sha`.

### Tests

- `test_head_sha_uses_api_order_not_timestamp` — verifies SHA selection uses topological order
- `test_not_stale_when_comment_after_all_timestamps` — regression test for the false-staleness edge case

### Known limitations

- `_get_pr_commits` fetches max 100 commits (no pagination). Tracked in #95 — extremely unlikely in stacked PR workflows.

Closes the staleness detection portion of #86.

### CODE REVIEW

**Comment by ichoosetoaccept** [comment]
* **#88** <a href="https://app.graphite.com/github/pr/detailobsessed/codereviewbuddy/88?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/detailobsessed/codereviewbuddy/88?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/codereviewbuddy/tools/comments.py [hunk 1]
```diff
@@ -426,9 +465,13 @@ async def list_review_comments(
...
+    commits = await call_sync_fn_in_threadpool(_get_pr_commits, owner, repo_name, pr_number, cwd=cwd)
 
-    threads = _parse_threads(raw_threads, pr_number, changed_files)
+    threads = _parse_threads(raw_threads, pr_number)
+
+    # Compute per-thread staleness (file changed after comment was posted)
+    await call_sync_fn_in_threadpool(_compute_staleness, threads, commits, owner, repo_name, cwd=cwd)
 
     # Include PR-level reviews from AI reviewers (e.g. Devin summaries)
     pr_reviews = await call_sync_fn_in_threadpool(_get_pr_reviews, owner, repo_name, pr_number, cwd=cwd)
```
[the remark this answers is not in the document]
Comment (ichoosetoaccept) [reply src/codereviewbuddy/tools/comments.py]: Correct — staleness is only meaningful for inline file comments. PR-level reviews and bot comments have no file context.

File: src/codereviewbuddy/tools/comments.py [hunk 2]
```diff
@@ -252,47 +229,109 @@ def _get_pr_issue_comments(
...
+
+        # No commits after the review → not stale
+        if review_point_sha == head_sha:
+            continue
+
+        # Fetch (cached) files changed since review point
+        if review_point_sha not in compare_cache:
+            compare_cache[review_point_sha] = _get_files_changed_between(owner, repo, review_point_sha, head_sha, cwd=cwd)
+
+        thread.is_stale = thread.file in compare_cache[review_point_sha]
```
[the remark this answers is not in the document]
Comment (ichoosetoaccept) [reply src/codereviewbuddy/tools/comments.py]: By design — the local `compare_cache` avoids repeated dict hashing in `gh.rest`'s cache layer and is scoped to a single `_compute_staleness` call. Both layers serve different purposes.

File: src/codereviewbuddy/tools/comments.py [hunk 3]
```diff
@@ -252,47 +229,116 @@ def _get_pr_issue_comments(
...
+    repo: str,
+    base_sha: str,
+    head_sha: str,
+    cwd: str | None = None,
+) -> set[str]:
+    """Get files changed between two commits using the compare API."""
+    result = gh.rest(f"/repos/{owner}/{repo}/compare/{base_sha}...{head_sha}", cwd=cwd)
+    if not result:
+        return set()
+    return {f["filename"] for f in result.get("files", []) if f.get("filename")}
```
[the remark this answers is not in the document]
Comment (ichoosetoaccept) [reply src/codereviewbuddy/tools/comments.py]: Fixed — `_get_files_changed_between` now catches `GhError` and returns an empty set instead of propagating the exception. This makes staleness detection resilient to force-push scenarios where the compare SHA no longer exists.

File: tests/test_comments.py [hunk 4]
```diff
@@ -153,11 +132,112 @@ def test_null_author_does_not_crash(self):
...
+        # Compare must target head_sha=bbb (topological), not aaa.
+        mock_rest = mocker.patch(
+            "codereviewbuddy.tools.comments.gh.rest",
+            return_value={"files": [{"filename": "src/codereviewbuddy/gh.py"}]},
+        )
+        _compute_staleness(threads, commits, "owner", "repo")
+        # Should compare bbb...bbb (review_point...head), both are bbb
+        mock_rest.assert_called_once()
+        call_url = mock_rest.call_args.args[0]
+        assert "bbb...bbb" in call_url
```
[the remark this answers is not in the document]
Comment (ichoosetoaccept) [reply tests/test_comments.py]: Good observation — the `bbb...bbb` self-compare would return empty files in production, so the thread wouldn't actually be marked stale in real usage. The test's purpose is to verify the SHA selection logic (that `head_sha` comes from API order, not timestamp order), not to test a realistic staleness outcome. The mock returning files is just to confirm the code path reaches the compare call with the correct SHAs.
